File: src/decoders/restaurant.ts
import type { Restaurant } from "~/models";
import { decodeMenu } from "./menu";
import { decodePaymentMethod } from "./payment-method";
import { decodeRestaurantKind } from "./restaurant-kind";
import { decodeContact } from "./contact";
import { decodeImage } from "./image";

export const decodeRestaurant = (restaurant: any): Restaurant => {
  return {
    id: restaurant.id,
    title: restaurant.title,
    latitude: restaurant.lat,
    longitude: restaurant.lon,
    area: restaurant.area,
    address: restaurant.adresse,
    opening: restaurant.opening,
    closing: restaurant.closing,
    kind: decodeRestaurantKind(restaurant.type),
    accessibility: restaurant.accessibility,
    wifi: restaurant.wifi,
    shortDescription: restaurant.shortdesc,
    description: restaurant.description,
    access: restaurant.access,
    operationalHours: restaurant.operationalhours,
    contact: decodeContact(restaurant.contact),
    crousAndGo: restaurant.crousandgo,
    album: restaurant.album && decodeImage(restaurant.album),
    photo: decodeImage(restaurant.photo),
    paymentMethods: restaurant.payment.map(decodePaymentMethod),
    menus: restaurant.menus.map(decodeMenu)
  };
};

File: library/src/commonMain/kotlin/ink/literate/crowous/decoders/restaurant.kt
package ink.literate.crowous.decoders

import kotlinx.serialization.json.*
import ink.literate.crowous.models.Restaurant

fun decodeRestaurant (restaurant: JsonObject): Restaurant {
  return Restaurant(
    id = restaurant["id"]!!.jsonPrimitive.int,
    title = restaurant["title"]!!.jsonPrimitive.content,
    latitude = restaurant["lat"]!!.jsonPrimitive.double,
    longitude = restaurant["lon"]!!.jsonPrimitive.double,
    area = restaurant["area"]!!.jsonPrimitive.content,
    address = restaurant["adresse"]!!.jsonPrimitive.content,
    opening = restaurant["opening"]!!.jsonPrimitive.content,
    closing = restaurant["closing"]!!.jsonPrimitive.content,
    kind = decodeRestaurantKind(restaurant["type"]!!.jsonPrimitive.content),
    accessibility = restaurant["accessibility"]!!.jsonPrimitive.boolean,
    wifi = restaurant["wifi"]!!.jsonPrimitive.boolean,
    shortDescription = restaurant["shortdesc"]!!.jsonPrimitive.content,
    description = restaurant["description"]!!.jsonPrimitive.content,
    access = restaurant["access"]!!.jsonPrimitive.content,
    operationalHours = restaurant["operationalhours"]!!.jsonPrimitive.content,
    contact = decodeContact(restaurant["contact"]!!.jsonObject),
    crousAndGo = restaurant["crousandgo"]!!.jsonPrimitive.content,
    album = restaurant["album"]?.let { decodeImage(it.jsonObject) },
    photo = decodeImage(restaurant["photo"]!!.jsonObject),
    paymentMethods = restaurant["payment"]!!.jsonArray.map {
      decodePaymentMethod(it.jsonObject)
    },
    menus = restaurant["menus"]!!.jsonArray.map {
      decodeMenu(it.jsonObject)
    }
  )
}
